R/MSnExperiment-class.R

```diff
@@ -142,16 +221,114 @@ readMSnExperiment <- function(file, sampleData, backend = BackendMzR(),
                           sep = "\n")) ## see issue #160
         hdr[order(hdr$acquisitionNum), ]
     }
-    spectraData <- DataFrame(do.call(rbind, bplapply(
-        file, .read_file, files=file, smoothed=smoothed, BPPARAM=BPPARAM
-    )))
+    spectraData <- DataFrame(
+        do.call(rbind, bplapply(file, .read_file, files=file,
+                                smoothed=smoothed, BPPARAM=BPPARAM)))
     backend <- backendInitialize(backend, file, spectraData, BPPARAM=BPPARAM)
-    backend <- backendImportData(backend, file, spectraData, BPPARAM=BPPARAM)
+    backend <- backendImportData(backend, spectraData, BPPARAM=BPPARAM)
     new("MSnExperiment",
         backend = backend,
         sampleData = sampleData,
         spectraData = spectraData,
-        processing = paste0("Data loaded [", date(), "]"),
-        files = file
+        processingQueue = list(),
+        processing = paste0("Data loaded [", date(), "]")
     )
 }
+
+#' @rdname MSnExperiment
+setMethod("spectrapply", "MSnExperiment", function(object, FUN = NULL,
+                                                   BPPARAM = bpparam(), ...) {
+    BPPARAM <- getBpParam(object, BPPARAM = BPPARAM)
+    isOK <- validateFeatureDataForOnDiskMSnExp(object@spectraData)
+    if (length(isOK))
+        stop(isOK)
+    file_f <- factor(object@spectraData$fileIdx,
```

Comment:
What do you think about an `fcol` argument? (see also https://github.com/lgatto/MSnbase/issues/422#issuecomment-459114666)

Reply:
I like the principal idea. To keep it even more generic I would however then go for an argument `f = spectraData(object)$fileIdx`. This gives the user even more possibilities to define the split, without requiring the column to be in the `spectraData`. If we do that we **have** to ensure that the returned spectra are in the expected order!